src/components/layout/index.js

```diff
@@ -0,0 +1,64 @@
+import React, {createContext} from 'react';
+import cx from 'classnames';
+import PropTypes from 'prop-types';
+
+const LayoutContext = createContext({
+  isSidebarVisible: true
+});
+
+export class Layout extends React.Component {
+  static displayName = 'Layout';
+
+  static propTypes = {
+    children: PropTypes.node,
+  }
+
+  state = {
+    isSidebarVisible: false
+  };
+
+  render() {
+    return (
+      <LayoutContext.Provider
+        value={ {
+          showSidebar: () => this.setState({ isSidebarVisible: true }),
+          hideSidebar: () => this.setState({ isSidebarVisible: false }),
+          isSidebarVisible: this.state.isSidebarVisible,
+        } }
+      >
+        <LayoutContext.Consumer>
+          {(context) => (
+            <div
+              className={ cx('layout-root', {
+                'sidebar-visible': context.isSidebarVisible,
+                'sidebar-hidden': !context.isSidebarVisible
+              }) }
+            >
+              { this.props.children }
+            </div>
+          )}
+        </LayoutContext.Consumer>
+      </LayoutContext.Provider>
+    );
+  }
+}
+
+
+export const Sidebar = ({ children }) => (
+  <aside className="sidebar">
+    <nav className="sidebar-navigation">
+      { children }
+    </nav>
+  </aside>
+);
+
+Sidebar.displayName = 'Sidebar';
+
+Sidebar.propTypes = {
+  children: PropTypes.node,
+};
+
+export default {
+  Provider: Layout,
```

Comment:
We should not export it in this way. We should to export the layout itself to let others to render it but we don't have to call it `Provider`. It's the layout component which has control over its children ui components and yes it provides methods to control main layout elements (like the sidebar) remotely but for those who don't want to use these methods it can be confusing why it is called Provider. On the other hand we want to expose the aforementioned methods to make grandchildren reach them but we only use the context's provider element within this component. Only the consumer element will be used outside.

So I can image something like this:

```
export default Layout;

export LayoutContext;
```